### ISSUE TITLE
Add quota support for storage backends

### ISSUE THREAD

**Issue by fjakobs** [docker/docker#3804]
I understand that the aufs does not support quotas but since the device mapper backend mounts an ext4 partition it should be possible to enable quota support there. If the loopback device with the ext4 partition is mounted with the proper quota options I think it would be possible to use regular linux quota tools inside of a docker container.

**Comment by nfnty** [comment]
@majst01 Any updates on the btrfs quota support? Very interested in this feature.

**Comment by blackfader** [comment]
@majst01 Any updates on the btrfs quota support? Very interested in this feature. +1

**Comment by denghongcai** [comment]
Any updates?

**Comment by SunWeicheng0001** [comment]
Any updates?

**Comment by xiaods** [comment]
Any updates?

**Comment by yijia2413** [comment]
ok now ?

**Comment by cpuguy83** [comment]
Docker 1.8 supports blkio-weight.
#15879 adds blkio iops per device
#14466 adds read/write bps per device
#13959 adds blkio weight per device

**Comment by denghongcai** [comment]
@cpuguy83 feature you listed here is not related with disk space quota...

**Comment by cpuguy83** [comment]
They are not "disk quota support", but they are related to the topic.

**Comment by kovacsa91** [comment]
I have read several forum topics and also issues in this topic, but haven't find exact, working solution for this problem. Tried to use DeviceMapper storage plugin, but don't know exactly, how to use it. Maybe is there any 3rd party solution for this?

**Comment by beetree** [comment]
It seems that this thread is concluding that this is NOT possible. But, there are arguments to the daemon that suggests it should be possible (https://github.com/snitm/docker/blob/master/daemon/graphdriver/devmapper/README.md).

Is the firm conclusion still that limiting disk usage _is not possible_ as of 1.9.1?

**Comment by vote539** [comment]
+1 for adding disk quotas

**Comment by tjdett** [comment]
@beetree As of 1.9.1, there is no support for limiting disk usage as a direct option. The `devicemapper` storage driver is implemented as a block device (and filesystem) per container. This limits disk usage, but only as a daemon-wide size limit for all new containers, and only while using that specific storage driver. It is the only Docker-base mechanism for limiting container size of which I am aware.

As discussed previously in this thread, ZFS and BTRFS are implemented in such a way that it is technically possible to add a quota to the container filesystem after it is created. If you were interested in implementing this, I would recommend writing a daemon to monitor the Docker event API and trigger the required quota change (potentially based on container metadata). Again, this is somewhat adhoc, but it's better than nothing.

Volume quota is a topic for a different thread, but it appears unlikely to become a core feature now that volume plugins exist.

**Comment by vote539** [comment]
As it currently stands (and correct me if I'm wrong), I believe that a Docker container, being able to write a large amount of data to disk, is capable of consuming the entire capacity of the host's file system.  This means that a container can cause "no space left on device" (ENOSPC) errors to other processes and containers running on the same host.

A design principle of Docker is to prevent containerized processes from causing I/O errors.  From a UX point of view, why should handling disk space limits be any different from handling memory or CPU limits?

**Comment by cpuguy83** [comment]
@vote539 You can certainly set block-level read/write controls, but these do not apply to the container's rootfs.
Theoretically one should be able to pass on quotas to both btrfs and zfs, so there could be support added here for these two filesystems only.

devicemapper will already create a fixed size device for each container's rootfs, so these can be limited in that way. Support in Docker already exists for changing these values at the daemon level.

**Comment by fahimeh2010** [comment]
any update in disk quota (not daemon-wide limitation)?

**Comment by zhuguihua** [comment]
@fahimeh2010 I sent a PR (#19651)about btrfs disk quota, but it must have more discussion on ux. You can go to see it, and welcome more suggestions.

**Comment by denghongcai** [comment]
@zhuguihua well done! hope it will merge into mainline soon!

**Comment by arkkanoid** [comment]
So, it's not possible to use Disk Quotas inside a contanier for now?

**Comment by cpuguy83** [comment]
@arkkanoid correct

**Comment by arkkanoid** [comment]
And is there another way to limit the disk size for an specific user inside docker container?

**Comment by vote539** [comment]
The solution I ended up using, which is far from perfect but gets the job done, is to make a limited-size file system loop-mounted on the host, and then mount it as a volume on the container.  This page has a good tutorial on setting up limited-size filesystems (the quota reporting mechanisms are optional):

http://souptonuts.sourceforge.net/quota_tutorial.html

One downside of this approach is that you need to have root permissions in order to mount the loop filesystem.  (That's not really a problem though because you also need root permissions to control Docker. :wink: )  The other downside is that only the one directory is going to be quota'd.  If the user inside the container tries to write a file outside of their designated directory, they can still take up space on the host filesystem.  Depending on your setup, read/write permissions inside the container could help deal with that potential problem.

**Comment by tjdett** [comment]
@vote539 This is essentially what the `devicemapper` storage backend does when used in loopback mode. To quote @jasonbrooks from Project Atomic:

[Friends Don't Let Friends Run Docker on Loopback in Production](http://www.projectatomic.io/blog/2015/06/notes-on-fedora-centos-and-docker-storage-drivers/)

If this is just for dev, then stop reading now. You have a solution that works, just go with it.

If it's for prod, then you will probably get better performance from using:
 * an LVM logical volume, or
 * an LVM thin pool volume, or
 * normal disk partitions, or
 * a single BTRFS/ZFS filesystem based on one of the above, using subvolumes/datasets to limit directory space.

You should also consider if `devicemapper` would be an easier solution depending on the environment you're running Docker in.

**Comment by gegere** [comment]
Basically the solution for the Docker disk limit stuck at 20gb would be to share the growing file directory to the host system? 

https://docs.docker.com/compose/compose-file/#volumes-volume-driver

```
    volumes:
   - "~/SITES/docker/mysql:/var/lib/mysql"
```

**Comment by tjdett** [comment]
@gegre You should store your data in a volume. That part you've definitely got right. The rest isn't good practice, and seems completely unrelated to the topic of this issue.

AFAIK, the only Docker storage backend which limits container disk by default is `devicemapper`. You can configure the daemon to adjust that size, so it's not "stuck" in any real sense.

Volumes use flat storage rather than layered storage, and use a completely different backend which can be extended via plugins. Simply making `/var/lib/mysql` a volume would ordinarily remove the 20GB limit from `devicemapper`, and also improve performance.

You do not need to map it to the host filesystem as you're suggesting - this is a pre-1.9 way of doing things, and was discouraged even then in favour of _data only containers_. Now that _named volumes_ exist, you should use them for your data instead.

**Comment by cpuguy83** [comment]
FYI, quotas are now supported in devicemapper, zfs, and btrfs.
Probably won't ever be supported in aufs or overlay.

**Comment by mYmNeo** [comment]
Although disk quota is supported by ext4, but at most of time the user running in the container is root, it is not limited by disk quota

**Comment by treeder** [comment]
@cpuguy83 do you have links to any information on this?

**Comment by cpuguy83** [comment]
@treeder Looks like docs are not ready for that yet, though there is some mention in `docs/reference/commandline/dockerd.md`

Basically, can set `--storage-opt size=1G` at either the daemon level or at runtime.

**Comment by tobegit3hub** [comment]
This is closed but I'm not sure if it works for our requirements. 

Does this work for btrfs, devicemapper or any storage backend? Which version of docker daemon has merged the patch and we can start to use?

**Comment by kovacsa91** [comment]
I think it isn’t working yet. I haven’t read that this feature already
implemented.

tobe <notifications@github.com> ezt írta (időpont: 2016. aug. 25., Cs 5:50):

**Comment by AkihiroSuda** [comment]
@tobegit3hub @kovacsa91 
PRs are available:
XFS+overlay2: https://github.com/docker/docker/pull/24771
XFS+overlay: https://github.com/docker/docker/pull/24807

**Comment by tobegit3hub** [comment]
Thanks @kovacsa91 and @AkihiroSuda . That's clear to me now 😃

**Comment by bklau** [comment]
@AkihiroSuda What is --storage-opt size=xxx means for the various storage drivers?.  Is  'xxx' means just the max size of the writable(CoW) layer in the container or the sum total of the container base image layers plus the writable(CoW) layer?. It's not clear to me. Also, what this "disk quota" field in the container.HostConfig?. How it is related to --storage-opt size=XXX ?

**Comment by timchenxiaoyu** [comment]
if I use devicemapper /aufs /overlay ...,all support  storage-opt size?

**Comment by Soullivaneuh** [comment]
This is not supported for `aufs` either `vfs`:

```
$ docker create -it --storage-opt size=5G debian /bin/bash
Error response from daemon: --storage-opt is not supported for aufs
(reverse-i-search)` sh': git^Chow HEAD
```

```
/ # docker create -it --storage-opt size=5G debian /bin/bash
Error response from daemon: --storage-opt is not supported for vfs
```

The issue should be re-opened.

**Comment by katsar0v** [comment]
Any updates on this issue?

**Comment by AkihiroSuda** [comment]
@katsar0v https://github.com/moby/moby/pull/24771

**Comment by geertschuring** [comment]
I get this error:

> docker: Error response from daemon: --storage-opt is supported only for overlay over xfs with 'pquota' mount option.

Im quite surpised that setting a max on file storage is not supported by docker.

**Comment by cpuguy83** [comment]
@geertschuring Setting quotas is highly dependent on the underlying filesystem.

Quotas are support on btrfs, overlay2+xfs, and and devicemapper.

**Comment by samip5** [comment]
The issue indeed should be re-opened.

```
$ docker create --storage-opt size=500G -v /opt/docker/appdata/nextcloud:/var/www/html -v /opt/docker/appdata/nextcloud/custom_apps:/var/www/html/custom_apps -v /opt/docker/appdata/nextcloud/config:/var/www/html/config -v /opt/docker/appdata/nextcloud/data:/var/www/html/data nextcloud:apache nextcloud

Error response from daemon: --storage-opt is supported only for overlay over xfs with 'pquota' mount option
```

**Comment by cpuguy83** [comment]
@samip5 unless there is some generic tech to target, then no.
Quotas are supported when the underlying storage supports it.

**Comment by samip5** [comment]
@cpuguy83 To my understanding, ext4 is supposed to support quotas, so why is it not supported by Docker?

**Comment by cpuguy83** [comment]
Seems like you are looking for https://github.com/moby/moby/issues/29364


Brian Goff

**Comment by andres-fr** [comment]
> @samip5 unless there is some generic tech to target, then no.
> Quotas are supported when the underlying storage supports it.

I don't think that is a sufficient argument. By that token, we wouldn't have cross-platform functionalities on the first place right?

**Comment by cpuguy83** [comment]
@andres-fr I'm not sure what you are getting at.
Docker supports quotas, but it is totally dependent on the underlying filesystem implementation.

For instance, with xfs you need the `pquota` mount option.
These are abstracted away as the `--storage-opt size=<val>` option.

**Comment by andres-fr** [comment]
TLDR: I think I'm at the wrong issue, sorry for the noise.

--------------------

To give more context, I was under the impression that this issue was about allowing the Docker host to have a disk usage limit (e.g. something in the form images+containers+volumes <= 100GB). After some searching I came to the conclusion that this functionality is not provided by Docker, and that the reason you provide is backend heterogeneity. Maybe at this point I dare to ask, do you know if this is this possible now?

**Comment by cpuguy83** [comment]
Docker will not impose such a limit.
You can put docker's root onto a separate disk/partition/quota.
